README.md
# shadowsocks-go

Current version: 1.2.2 [![Build Status](https://travis-ci.org/shadowsocks/shadowsocks-go.png?branch=master)](https://travis-ci.org/shadowsocks/shadowsocks-go)

shadowsocks-go is a lightweight tunnel proxy which can help you get through firewalls. It is a port of [shadowsocks](https://github.com/clowwindy/shadowsocks).

The protocol is compatible with the origin shadowsocks (if both have been upgraded to the latest version).

**Note `server_password` option syntax changed in 0.6.2, the client now connects to servers in the order specified in the config.**

**Please develop on the latest develop branch if you want to send pull request.**

# Install

Download precompiled binarys from the [release page](https://github.com/shadowsocks/shadowsocks-go/releases). (All compiled with cgo disabled, except the mac version.)

You can also install from source (assume you have go installed):

```
# on server
go get github.com/shadowsocks/shadowsocks-go/cmd/shadowsocks-server
# on client
go get github.com/shadowsocks/shadowsocks-go/cmd/shadowsocks-local
```

It's recommended to disable cgo when compiling shadowsocks-go. This will prevent the go runtime from creating too many threads for dns lookup.

# Usage

Both the server and client program will look for `config.json` in the current directory. You can use `-c` option to specify another configuration file.

Configuration file is in json format and has the same syntax with [shadowsocks-nodejs](https://github.com/clowwindy/shadowsocks-nodejs/). You can download the sample [`config.json`](https://github.com/shadowsocks/shadowsocks-go/blob/master/config.json), change the following values:

```
server          your server ip or hostname
server_port     server port
local_port      local socks5 proxy port
method          encryption method, null by default (table), the following methods are supported:
                    aes-128-cfb, aes-192-cfb, aes-256-cfb, bf-cfb, cast5-cfb, des-cfb, rc4-md5, rc4-md5-6, chacha20, salsa20, rc4, table
password        a password used to encrypt transfer
timeout         server option, in seconds
```

Run `shadowsocks-server` on your server. To run it in the background, run `shadowsocks-server > log &`.

On client, run `shadowsocks-local`. Change proxy settings of your browser to

```
SOCKS5 127.0.0.1:local_port
```

## About encryption methods

AES is recommended for shadowsocks-go. [Intel AES Instruction Set](http://en.wikipedia.org/wiki/AES_instruction_set) will be used if available and can make encryption/decryption very fast. To be more specific, **`aes-128-cfb` is recommended as it is faster and [secure enough](https://www.schneier.com/blog/archives/2009/07/another_new_aes.html)**.

**rc4 and table encryption methods are deprecated because they are not secure.**

### One Time Auth

OTA function is deprecated because it is reported to have potential security risk.

## Command line options

Command line options can override settings from configuration files. Use `-h` option to see all available options.

```
shadowsocks-local -s server_address -p server_port -k password
    -m aes-128-cfb -c config.json
    -b local_address -l local_port
shadowsocks-server -p server_port -k password
    -m aes-128-cfb -c config.json
    -t timeout
```

Use `-d` option to enable debug message.

## Use multiple servers on client

```
server_password    specify multiple server and password, server should be in the form of host:port
```

Here's a sample configuration [`client-multi-server.json`](https://github.com/shadowsocks/shadowsocks-go/blob/master/sample-config/client-multi-server.json). Given `server_password`, client program will ignore `server_port`, `server` and `password` options.

Servers are chosen in the order specified in the config. If a server can't be connected (connection failure), the client will try the next one. (Client will retry failed server with some probability to discover server recovery.)

## Multiple users with different passwords on server

The server can support users with different passwords. Each user will be served by a unique port. Use the following options on the server for such setup:

```
port_password   specify multiple ports and passwords to support multiple users
```

Here's a sample configuration [`server-multi-port.json`](https://github.com/shadowsocks/shadowsocks-go/blob/master/sample-config/server-multi-port.json). Given `port_password`, server program will ignore `server_port` and `password` options.

### Update port password for a running server

Edit the config file used to start the server, then send `SIGHUP` to the server process.

# Note to OpenVZ users

**Use OpenVZ VM that supports vswap**. Otherwise, the OS will incorrectly account much more memory than actually used. shadowsocks-go on OpenVZ VM with vswap takes about 3MB memory after startup. (Refer to [this issue](https://github.com/shadowsocks/shadowsocks-go/issues/3) for more details.)

If vswap is not an option and memory usage is a problem for you, try [shadowsocks-libev](https://github.com/madeye/shadowsocks-libev).


CHANGELOG
1.2.2 (2018-06-10)
    * OTA is finally deprecated.
    * Codes clear up.

1.2.1 (2017-01-22)
    * Fix bug in handling of UDP OTA packets

1.2.0 (2017-01-20)
    * Support UDP relay on server side, and OTA
    * Support "aes-[128/192/256]-ctr" encryption method (Thanks for @slurin)
    * Support "chacha20-ietf" encryption method
    * Improve performance of "chacha20" encryption method
    * Correctly close connection if handshake failed

1.1.5 (2016-05-04)
    * Support OTA (Thanks for @ayanamist for implementing this feature)

1.1.4 (2015-05-10)
    * Support "chacha20" encryption method, thanks to @defia
    * Support "salsa20" encryption method, thanks to @genzj
    * Fix go 1.4 canonical import paths, thanks to @ddatsh
    * Exit if port not bindable, thanks to @thomasf
    * Always set timeout for data transfer
    * More buffer reuse

1.1.3 (2014-09-28)
    * Fix can't specify encryption method in config file

1.1.2 (2014-09-21)
    * Support new encryption method "rc4-md5"
    * Use aes-256-cfb as default encryption method for command line app

1.1.1 (2013-07-12)
    * Add -b option to limit listen address for client
    * Fix can't override server address on command line

1.1 (2013-05-26)
    * Add more encryption methods
    * Enable CGO for OS X when building

1.0 (2013-05-17)
    * Support specify servers with IPv6 address
    * Support IPv6 address in socks requests
    * Better handling of configuration file for debian package

0.6.2 (2013-03-15)
    * Connect to multiple servers in the order specified in config
    * More information in server error log to help error diagnosing

0.6.1 (2013-03-04)
    * Small performance improvement
    * For windows: provide shadowsocks-tray.exe to hide shadowsocks-local.exe in
      system tray. (Thanks to phuslu's taskbar.)

0.6 (2013-01-23)

    * Generate ciphers on demand (encryption table cache is removed)
    * Support RC4 encryption


cmd/cipher_ss.go
package main

import (
	ss "github.com/qunxyz/shadowsocks-go/shadowsocks"
	"crypto/cipher"
	"fmt"
)

type cipherInfo struct {
	method string
	KeySize    int
	IVSize     int
}

var cipherMethod = map[string]*cipherInfo{
	"aes-128-cfb":   {"aes-128-cfb", 16, 16},
	"aes-192-cfb":   {"aes-192-cfb", 24, 16},
	"aes-256-cfb":   {"aes-256-cfb", 32, 16},
	"aes-128-ctr":   {"aes-128-ctr", 16, 16},
	"aes-192-ctr":   {"aes-192-ctr", 24, 16},
	"aes-256-ctr":   {"aes-256-ctr", 32, 16},
	"des-cfb":       {"des-cfb", 8, 8},
	"bf-cfb":        {"bf-cfb", 16, 8},
	"cast5-cfb":     {"cast5-cfb", 16, 8},
	"rc4-md5":       {"rc4-md5", 16, 16},
	"chacha20":      {"chacha20", 32, 8},
	"chacha20-ietf": {"chacha20-ietf", 32, 12},
	"chacha20-ietf-poly1305": {"chacha20-ietf-poly1305", 32, 32},
	"salsa20":       {"salsa20", 32, 8},
}

func check_cipher(info *cipherInfo, src []byte) {
	c, err := ss.NewCipher(info.method, "123456"); if err != nil { fmt.Print(err) }

	if info.method == "chacha20-ietf-poly1305" {
		///////////////////////////////////////
		fmt.Println("----------------------------------------")
		fmt.Printf("Testing aead %s\n", info.method)
		fmt.Println("----------------------------------------")

		iv, err := c.NewIV(); if err != nil { fmt.Print(err) }
		if err = c.Init(iv, ss.Encrypt); err != nil { fmt.Print(err) }

		dst := make([]byte, c.GetCryptor(ss.Encrypt).(cipher.AEAD).Overhead()+len(src))
		err = c.Encrypt(dst, src); if err != nil { fmt.Print(err) }

		fmt.Println("check data encrypted")
		fmt.Printf("data: %d(%d)\n", dst, len(dst))
		fmt.Printf("iv: %d\n", iv)
		fmt.Println("----------------------------------------")

		err = c.Init(iv, ss.Decrypt); if err != nil { fmt.Print(err) }

		src = dst
		dst = make([]byte, 2+c.GetCryptor(ss.Decrypt).(cipher.AEAD).Overhead()+len(src))

		err = c.Decrypt(dst, src); if err != nil { fmt.Print(err) }

		fmt.Println("check data decrypted")
		fmt.Printf("data: %s\n", dst)
		fmt.Printf("iv: %d\n", iv)
		fmt.Println("----------------------------------------")
	} else {
		///////////////////////////////////////
		fmt.Printf("Testing stream %s\n", info.method)
		fmt.Println("----------------------------------------")

		dst := make([]byte, len(src))
		iv, err := c.NewIV(); if err != nil { fmt.Print(err) }

		err = c.Init(iv, ss.Encrypt); if err != nil { fmt.Print(err) }
		err = c.Encrypt(dst, src); if err != nil { fmt.Print(err) }
		fmt.Println("check data encrypted")
		fmt.Printf("data: %d\n", dst)

		err = c.Init(iv, ss.Decrypt); if err != nil { fmt.Print(err) }
		err = c.Decrypt(dst, dst); if err != nil { fmt.Print(err) }

		fmt.Println("----------------------------------------")
		fmt.Println("check data decrypted")
		fmt.Printf("data: %s\n", dst)
		fmt.Println("----------------------------------------")
		///////////////////////////////////////
	}
}

func main() {
	src := []byte("hello")
	for _, item := range cipherMethod {
		check_cipher(item, src)
	}
	//////////////////////////////////////////////////////////////
	///////////////////////////////////////
	fmt.Println("Testing stream")
	fmt.Println("----------------------------------------")
	c, err := ss.NewCipher("aes-128-cfb", "123456"); if err != nil { fmt.Print(err) }
	dst := make([]byte, len(src))
	iv, err := c.NewIV(); if err != nil { fmt.Print(err) }
	err = c.Init(iv, ss.Encrypt); if err != nil { fmt.Print(err) }
	err = c.Encrypt(dst, src); if err != nil { fmt.Print(err) }
	fmt.Println("check data encrypted")
	fmt.Printf("data: %d\n", dst)

	err = c.Init(iv, ss.Decrypt); if err != nil { fmt.Print(err) }
	err = c.Decrypt(dst, dst); if err != nil { fmt.Print(err) }

	fmt.Println("----------------------------------------")
	fmt.Println("check data decrypted")
	fmt.Printf("data: %s\n", dst)
	fmt.Println("----------------------------------------")
	///////////////////////////////////////
	fmt.Println("----------------------------------------")
	fmt.Println("Testing aead")
	fmt.Println("----------------------------------------")
	c, err = ss.NewCipher("chacha20-ietf-poly1305", "123456"); if err != nil { fmt.Print(err) }

	iv, err = c.NewIV(); if err != nil { fmt.Print(err) }
	err = c.Init(iv, ss.Encrypt); if err != nil { fmt.Print(err) }

	dst = make([]byte, c.GetCryptor(ss.Encrypt).(cipher.AEAD).Overhead()+len(src))
	err = c.Encrypt(dst, src); if err != nil { fmt.Print(err) }

	fmt.Println("check data encrypted")
	fmt.Printf("data: %d(%d)\n", dst, len(dst))
	fmt.Printf("iv: %d\n", iv)
	fmt.Println("----------------------------------------")

	err = c.Init(iv, ss.Decrypt); if err != nil { fmt.Print(err) }

	src = dst
	dst = make([]byte, 2+c.GetCryptor(ss.Decrypt).(cipher.AEAD).Overhead()+len(src))
	err = c.Decrypt(dst, src); if err != nil { fmt.Print(err) }

	fmt.Println("check data decrypted")
	fmt.Printf("data: %s\n", dst)
	fmt.Printf("iv: %d\n", iv)
	fmt.Println("----------------------------------------")
}